### DISCUSSION TITLE
Is there anyway to implement a Menu Bar inside the Title Bar?

### DISCUSSION THREAD

**Discussion by LucianoSaldivia** [TomSchimansky/CustomTkinter#1422] [Q&A]
In a professional looking app like Visual Studio Code, you'd have this menu bar:
![image](https://user-images.githubusercontent.com/32425456/229679821-2fda0872-ca80-4ada-b813-65217d1d3109.png)

implemented inside the title bar:
![image](https://user-images.githubusercontent.com/32425456/229680569-d44a97f0-a472-4d8f-b8e0-28b08d9164fa.png)

I've been searching everywhere, and while you can start with `overrideredirect(True)`, there are many issues when trying to implement something like this with basic widgets. 

No other GUI package has this ability.
If someone knows how to implement something as shown in the images, please let me know how.

**Comment by ElectricCandlelight** [comment]
Could try using on enter on a label that then places a frame below it with the menu items, then on leave for the frame to remove it.

**Reply by LucianoSaldivia** [reply]
I'm asuming you mean `"<Enter>"` and `"<Leave>"` events. I've tried a similar approach, but i don't want the whole menu to disappear once you leave a frame (or the submenus frames). The whole menu is supposed to disappear only when you select an option (which is easy to implement) and when you click elsewhere on the screen. The latter you can't implement, since tkinter only has `"<Button-1>"` event for when the user clicks _**inside**_ of the the widget, and i need an event for a click _**outside**_ the frame/s.

Also, i've seen pseudo-solutions such as creating an invisible frame behind the menu, so that when you click that frame the menu disappears, but then you need the root object to be accessible inside the dropdown menu, which you don't get, you just get the `master` which in this case should be the TitleBar. You can workaround this, but it's not a clean OOP solution, which is how it should be implemented.

**Comment by Akascape** [comment]
@LucianoSaldivia 
No, tkinter itself can't change title bar parameters except icons and title. But there are some ideas you can try:

Change the title bar color (guide is given here: https://github.com/TomSchimansky/CustomTkinter/discussions/1299)
Then you can add a narrow frame below the tile bar where you can place the buttons.

Or, you can recreate a new title bar with customtkinter by disabling the native title bar. Here is a full ready-made template: https://github.com/TomSchimansky/CustomTkinter/discussions/736

Other than that python can't change window attributes, you can only do that with C (or ctypes maybe incase of python), an example library is here: https://github.com/littlewhitecloud/CustomTkinterTitlebar (Don't know if it will work properly)

Edit: This menubar will work: https://github.com/TomSchimansky/CustomTkinter/discussions/1670

**Reply by littlewhitecloud** [reply]
LOL, it won't work correctly, the menubar will show below the `titlebar` but you can put `Menubutton` in the titlebar.

**Comment by LucianoSaldivia** [comment]
Hi, first of all, thanks for such a complete answer and also being so responsive!

I've tried the three approaches you mention, and the farthest i've gotten was with the second one, disabling the native title bar and creating my own. But the DropdownMenu is giving me lots of issues, that's the main reason i'm looking for another approach. 
Here's my code so far (some parts are in spanish, my mother language):

```python
import customtkinter as ctk
import tkinter as tk
from functools import partial
from typing import Callable



OPTION_BUTTON_PADX = 3
OPTION_BUTTON_PADY = 3




class _CDMOptionButton(ctk.CTkButton):
    def setParentMenu(self, menu: "CustomDropdownMenu"):
        self.parent_menu = menu

class _CDMSubmenuButton(_CDMOptionButton):
    def setSubmenu(self, submenu: "CustomDropdownMenu"):
        self.submenu = submenu

class CustomDropdownMenu(ctk.CTkFrame):
    def __init__(self, 
                 master, 
                 menu_seed_object: ctk.CTkBaseClass | _CDMSubmenuButton,
                 border_width: int = 1, 
                 fg_color = None, 
                 border_color: str | tuple[str, str] = "grey50", 

                 item_fg_color: str | tuple[str, str] = "transparent", 
                 item_hover_color: str | tuple[str, str] = "grey20", 
                 item_font: ctk.CTkFont = ("helvetica", 10),
                 ):
        super().__init__(
            master=master,
            border_width=border_width,
            fg_color=fg_color,
            border_color=border_color,
        )
        
        self.menu_seed_object = menu_seed_object
        self.item_fg_color = item_fg_color
        self.item_hover_color = item_hover_color
        self.item_font = item_font

        self._options_list: list[_CDMOptionButton | _CDMSubmenuButton] = []

        # Se dispara al entrar en cualquier opción?
        self.bind("<Leave>", self._checkIfMouseLeft)
        
        # hidden by default

    def selectOption(self, command) -> None:
        self._hideAllMenus()
        command()
    
    def addOption(self, option: str, command: Callable) -> None:
        optionButton = _CDMOptionButton(
            self, 
            text=option,
            anchor="w",
            command=partial(self.selectOption, command)
        )
        optionButton.setParentMenu(self)
        self._options_list.append(optionButton)
        self._configureButton(optionButton)

        optionButton.pack(
            side="top",
            fill="both", 
            expand=True,
            padx=OPTION_BUTTON_PADX,
            pady=OPTION_BUTTON_PADY,
        )
    def addSubmenuFromSeed(self, submenuSeedName: str) -> "CustomDropdownMenu":
        submenuButtonSeed = _CDMSubmenuButton(self, text=submenuSeedName, anchor="w")
        submenuButtonSeed.setParentMenu(self)
        self._options_list.append(submenuButtonSeed)
        self._configureButton(submenuButtonSeed)

        submenu = CustomDropdownMenu(
            master=self.master,
            menu_seed_object=submenuButtonSeed,
            item_fg_color=self.item_fg_color,
            item_hover_color=self.item_hover_color,
            item_font=self.item_font,
        )
        submenuButtonSeed.setSubmenu(submenu=submenu)
        submenuButtonSeed.configure(command=submenu.toggleShow)

        submenuButtonSeed.bind("<Enter>", submenu._show)

        submenuButtonSeed.pack(
            side="top",
            fill="both", 
            expand=True,
            padx=3,
            pady=3,
        )
        return submenu
    def addSeparator(self) -> None:
        # separator = ttk.Separator(self, orient="horizontal", bg=self._border_color) # No puedo cambiar el color
        separator = ctk.CTkFrame(
            master=self, 
            height=2, # No se puede hacer de 1px, es de mínimo 2px
            fg_color='grey20', 
            # fg_color=self._border_color, 
            border_width=0
        )
        separator.pack(
            side="top",
            fill="x", 
            expand=True,
        )

    def _show(self, *args, **kwargs) -> None:
        if isinstance(self.menu_seed_object, _CDMSubmenuButton):
            # Submenu
            self.place(
                in_=self.menu_seed_object.parent_menu,
                x=self.menu_seed_object.winfo_x() + self.menu_seed_object.winfo_width() + OPTION_BUTTON_PADX - 1,
                y=self.menu_seed_object.winfo_y() - OPTION_BUTTON_PADY,
            )
        else:
            # First Menu
            self.place(
                x=self.menu_seed_object.winfo_x(),
                y=self.menu_seed_object.winfo_y() + self.menu_seed_object.winfo_height(),
            )
        self.focus()    
    def _hide(self, *args, **kwargs) -> None:
        self.place_forget()
    def _hideParentMenus(self, *args, **kwargs) -> None:
        if isinstance(self.menu_seed_object, _CDMSubmenuButton):
            self.menu_seed_object.parent_menu._hideParentMenus()
            self.menu_seed_object.parent_menu._hide()
    def _hideChildrenMenus(self, *args, **kwargs) -> None:
        # Si self contiene submenú , lo cierro
        if any(isinstance(option, _CDMSubmenuButton) for option in self._options_list):
            for option in self._options_list:
                if isinstance(option, _CDMSubmenuButton):
                    option.submenu._hide()
    def _hideAllMenus(self, *args, **kwargs) -> None:
        self._hideChildrenMenus()
        self._hide()
        self._hideParentMenus()
    def _collapseSiblingSubmenus(self, button: _CDMOptionButton | _CDMSubmenuButton, *args, **kwargs) -> None:
        for option in self._options_list:
            if option != button and isinstance(option, _CDMSubmenuButton):
                option.submenu._hideChildrenMenus()
                option.submenu._hide()

    def toggleShow(self, *args, **kwargs) -> None:
        if not self.winfo_manager():
            self._show()
        else:
            self._hideChildrenMenus()
            self._hide()

    def _configureButton(self, boton: ctk.CTkButton) -> None:
        boton.configure(fg_color="transparent")
        if self.item_fg_color:
            boton.configure(fg_color=self.item_fg_color)
        if self.item_hover_color:
            boton.configure(hover_color=self.item_hover_color)
        if self.item_font:
            boton.configure(font=self.item_font)

        boton.bind("<Enter>", partial(self._collapseSiblingSubmenus, boton))
    def _getSubMenus(self) -> list["CustomDropdownMenu"]:
        if any(isinstance(option, _CDMSubmenuButton) for option in self._options_list):
            subMenusList = list()
            for option in self._options_list: 
                if isinstance(option, _CDMSubmenuButton):
                    subMenusList.append(option.submenu)
            return subMenusList
        else:
            return []

    def _contieneCoordenada(self, x_root, y_root) -> bool:
        return self.winfo_rootx() < x_root < self.winfo_rootx()+self.winfo_width() and \
            self.winfo_rooty() < y_root < self.winfo_rooty()+self.winfo_height()
    
    def _checkIfMouseLeft(self, event: tk.Event=None) -> None:
        # print(event.x_root, event.y_root, self._contieneCoordenada(event.x_root, event.y_root))
        # No está en el menu del evento
        if not self._contieneCoordenada(event.x_root, event.y_root):
            
            # Es Submenu y no está en el parent
            if isinstance(self.menu_seed_object, _CDMSubmenuButton) and not self.menu_seed_object.parent_menu._contieneCoordenada(event.x_root, event.y_root):

                # No tiene Submenus, o tiene y no está en ninguno de ellos
                subMenus = self._getSubMenus()
                if subMenus == [] or all((not submenu._contieneCoordenada(event.x_root, event.y_root)) for submenu in subMenus):
                    self._hideAllMenus()
            
            # No es submenú
            elif not isinstance(self.menu_seed_object, _CDMSubmenuButton):

                # No tiene Submenus, o tiene y no está en ninguno de ellos
                subMenus = self._getSubMenus()
                if subMenus == [] or all((not submenu._contieneCoordenada(event.x_root, event.y_root)) for submenu in subMenus):
                    self._hideAllMenus()




def dummy(*kwargs):
    print("dummy func called")


def main():
    root = ctk.CTk()
    root.geometry("800x500+200+200")

    boton = ctk.CTkButton(root, anchor="w")
    dropdownMenu = CustomDropdownMenu(master=root, menu_seed_object=boton)
    boton.configure(
        text="Archivo",
        corner_radius=7,
        fg_color="transparent",
        hover_color="darkblue",
        command=dropdownMenu.toggleShow,
    )
    dropdownMenu.addOption("Abrir...", dummy)
    dropdownMenu.addOption("Nuevo", dummy)
    dropdownMenu.addSeparator()
    exportarSubmenu = dropdownMenu.addSubmenuFromSeed(submenuSeedName="Exportar...")
    exportarSubmenu.addOption("Como PNG...", dummy)
    exportarSubmenu.addOption("Como PDF...", dummy)
    exportarSubmenu.addSeparator()
    otraCosaSubmenu = exportarSubmenu.addSubmenuFromSeed(submenuSeedName="Como otra cosa...")
    otraCosaSubmenu.addOption("Como querés exportarlo campeón?...", dummy)

    # boton.pack()
    boton.place(
        x=0,
        y=0,
    )
    root.mainloop()



if __name__ == "__main__":
    main()
```

As you can see, the whole menu closes either when you choose an option, or when you `"<Leave>"` all menus. Problem is, i want the second condition to be when you click somewhere else instead of `"<Leave>"`, andi can't find a proper way of doing this. As i said before, you can workaround the problem calling master to get the `"<Button-1>"` event, but that would only work when master is root, otherwise clicks wouldn't be detected properly.

**Comment by Akascape** [comment]
@LucianoSaldivia I just found a hacky way to make this possible in python:

![Screenshot](https://user-images.githubusercontent.com/89206401/229859251-cee58e53-6969-436b-ad11-db0fdba2d949.jpg)

```python
import customtkinter
from ctypes import windll, byref, sizeof, c_int

class CTkTitleMenu(customtkinter.CTkToplevel):
    
    def change_dimension(self):
        self.geometry(f"{root.winfo_width()-125-self.x_offset}x{root.winfo_height()}+{root.winfo_x()+self.x_offset}+{root.winfo_y()+self.y_offset}")
        self.deiconify()
            
    def change_header_color(self, caption_color):
        try:
            # optional feature to change the header color in windows
            HWND = windll.user32.GetParent(self.master.winfo_id())
            DWMWA_CAPTION_COLOR = 35
            windll.dwmapi.DwmSetWindowAttribute(HWND, DWMWA_CAPTION_COLOR, byref(c_int(caption_color)), sizeof(c_int))
        except:
            pass
        
    def __init__(
        self,
        master,
        title_bar_color = "default", 
        x_offset: int = 40,
        y_offset: int = 6):
        
        super().__init__()

        self.after(10)
        self.master = master
        self.master.minsize(200,100)
        self.master.title("")
        self.overrideredirect(True)
        
        if title_bar_color=="default":
            if customtkinter.get_appearance_mode()=="Light":
                title_bar_color = 0x999999 # RGB order: 0xrrggbb
                
            else:
                title_bar_color = 0x303030 # RGB order: 0xrrggbb
                
        self.transparent_color = self._apply_appearance_mode(self._fg_color)
        self.attributes("-transparentcolor", self.transparent_color)
        self.resizable(True, True)
        self.config(background=self.transparent_color)
        self.caption_color = title_bar_color
        self.change_header_color(self.caption_color)
        self.x_offset = x_offset
        self.y_offset = y_offset
        
        self.master.bind("<Configure>", lambda _: self.change_dimension())
        self.master.bind("<FocusIn>", lambda _: self.deiconify())
        self.master.bind("<FocusOut>", lambda _: self.withdraw() if self.master.state()=="iconic" else None)
        self.bind("<FocusIn>", lambda _: root.focus())

# Example Usage:

customtkinter.set_appearance_mode("Dark")

root = customtkinter.CTk()
header = CTkTitleMenu(root)

for i in range(0, 5):
    # add widgets in header frame
    customtkinter.CTkButton(header, text="Tab"+str(i), fg_color="grey15",
                            width=50, height=10).grid(row=0, column=i, padx=(0,10))

root.mainloop()
```

**Reply by LucianoSaldivia** [reply]
OMG This looks great! I don't have much time today and neither tomorrow, but will definitely look into it a bit more in a couple of days and will replay back. 
Again, thanks a lot, this is very useful!

**Reply by Akascape** [reply]
@LucianoSaldivia Can I use your `CustomDropdownMenu` class?

**Reply by LucianoSaldivia** [reply]
Of course! I took great inspiration from the ctk DropdownMenu class, if this is useful, feel free to use it for your own good and also (i'd be honoured) if applying some makeup it could become a ctk widget in the future as well feel free to do that too.

**Reply by Akascape** [reply]
@LucianoSaldivia Thanks, I am planning to make a new library for customtkinter combining my custom title bar menu and your dropdown menu.

**Reply by LucianoSaldivia** [reply]
Are you planning to make a whole new library or add packages to the customtkinter library for a future version?
Also, i'd be thankful if you throw my name in there, like in a comment inside the dropdown menu class file as a contributor or something like that, but it's ok if you don't hahaha
Finally, i have yet to implement the GUI for the _CDMSubmenuButton with a right-aligned right-pointing arrow, to indicate to the user that it's a Submenu and not just another option. I'm thinking about a button with two labels inside, but i'm not sure about it yet. If i get that to work i'll post it here.

**Reply by Akascape** [reply]
@LucianoSaldivia 
> Are you planning to make a whole new library or add packages to the customtkinter library for a future version?

Yes, a type of ctk add-on you can say.
I put contributor name in all of all projects, so don't worry. 

But now i think, my title bar menu is just a small hack with the toplevel window, I feel your dropdown menu deserve a place in the customtkinter library itself so that everyone can use it.

You should give a pull request in the near future.

**Reply by LucianoSaldivia** [reply]
About that contributor name, that's great then! I think your idea could be pretty useful, do you work alongside TomSchimansky, a group of devs or is this a one-dev work?
Have you published anything about this add-on or is it still in development?

I think i can polish the code a bit more before that, and i also think i should do my research on how to make a pull request, i have no idea how to do that. Thanks for the advide, i'll have that in mind.

**Reply by Akascape** [reply]
@LucianoSaldivia 
>  do you work alongside TomSchimansky, a group of devs or is this a one-dev work?

No, customtkinter is one-dev work only. But I like to build other useful add-ons for it in my free time.

> Have you published anything about this add-on or is it still in development?

As I said, I have not worked on it. It was just an idea that striked my mind. But your dropdown menu is a separate widget class.

So I feel you should give a pull request first, if TomSchimansky accepts it, then it will be good. 

Otherwise you can publish it as your own addon. Then everyone can use it in their projects.

**Reply by LucianoSaldivia** [reply]
Oh ok, i got it now. Thanks for everything!

**Reply by LucianoSaldivia** [reply]
Finally after some very full-of-work days, i could take a look into your code. I just have a couple of questions:
Why did you choose a CTkTopLevel and not something simpler to setup and configure such as a CTkFrame for example?
I'm experiencing a couple of buggy things. When i resize the window moving the mouse fast enough, you can see a black outline of the CTkTopLevel, very visible right at the left of the window control buttons (minimize, resize and close). I checked by chaging the fg_color to red, and yes, what i was noticing was the outline of this CTkTopLevel. 
Also, when maximizing the window, it looks as if it's not starting on (0,0) instead it starts about (-2,-2) you can see this by adding `print(self.geometry())` right after line 8 (`self.deiconify()`).
I tried giving the TopLevel a different geometry, to set it's height to the height of the titlebar but i don't want to hardcode the height value of it, and i can't seem to get it directly through winfo_ methods on root, i have no idea how to do this and it is necessary. I think if i can get this done, the buggy outline at resizing wouldn't be visible, therefore removing that problem.

I have to say, there's a *big pro* with this approach though, and it's the use of the default window control buttons. Otherwise i'd be forced to create my own minimize/resize/close buttons, and handling their functionality correctly is not easy to say the least.

**Reply by Akascape** [reply]
@LucianoSaldivia 

>Why did you choose a CTkTopLevel and not something simpler to setup and configure such as a CTkFrame for example?

Because CTkFrame can't be placed outside the root window. And the title bar part of the main window is actually considered outside. So only a toplevel can be placed in top of the header. 

> When i resize the window moving the mouse fast enough, you can see a black outline of the CTkTopLevel, very visible right at the left of the window control buttons (minimize, resize and close).

Actually I added `toplevel.resizable` method which disables the toplevel from displaying as an extra window in taskbar. If we remove this line then it will remove those blocky glitches but now the toplevel will behave as an active window in taskbar.

> Also, when maximizing the window, it looks as if it's not starting on (0,0) instead it starts about (-2,-2) you can see this by adding print(self.geometry()) right after line 8 (self.deiconify()).

That's the problem with tkinter, because the header height also becomes a little smaller when maximized. But it can be fixed by adjusting the toplevel y offset.

> I tried giving the TopLevel a different geometry, to set it's height to the height of the titlebar but i don't want to hardcode the height value of it...

Sure, it will remove those blocky glitches if you change the height of that toplevel (a fixed value), but in that case you can't place other widgets like your dropdown menu inside that toplevel. You have to spawn it in the root window only (which will look a little displaced). 

> I have to say, there's a big pro with this approach though, and it's the use of the default window control buttons. 

To be honest, there is no other way of implementing a title bar like that with tkinter.

Maybe there are some other methods to fix the toplevel window. 

Ig, you have to stay with some minor bugs...

> Otherwise i'd be forced to create my own minimize/resize/close buttons...

I tried making a full customised window: https://github.com/TomSchimansky/CustomTkinter/discussions/736

But it has lots of minor issues which can't be solved easily (issues like resiziable edges, no icons, no taskbar thumbnail...)

**Reply by YoussefEssalhi** [reply]
Hi @Akascape , this part of code not work in my `win 10` it's only work in my `win 11`  How can I fix it because I have already worked on a project for a long time and I need to change the color to dark blue [ `#031249` ] ?

```python
def change_header_color(self, caption_color):
    try:
        # optional feature to change the header color in windows
        HWND = windll.user32.GetParent(self.master.winfo_id())
        DWMWA_CAPTION_COLOR = 35
        windll.dwmapi.DwmSetWindowAttribute(HWND, DWMWA_CAPTION_COLOR, byref(c_int(caption_color)), sizeof(c_int))
    except:
        pass
```

my code that i use 

```python
try:
    from ctypes import windll, byref, sizeof, c_int
    HWND = windll.user32.GetParent(self.winfo_id())
    windll.dwmapi.DwmSetWindowAttribute(HWND, 35, byref(c_int(0x00491203)), sizeof(c_int))
except:
    pass
```

**Reply by Akascape** [reply]
@YoussefEssalhi it will not work in windows 10, that feature is only available for win11. I recommend you to use this package: https://github.com/Akascape/CTkMenuBar

**Reply by YoussefEssalhi** [reply]
@Akascape , I don't want to use `overrideredirect`, Is there any way to change the color in `Win 10`, I just only want to change the **color** ?
Thanks for your help .

**Reply by Akascape** [reply]
@YoussefEssalhi I can feel the pain but it is not possible to change the header color so easily in windows 10. You have to redraw everything which is really hard to do. 

I even made a ready made package that can change themes and header colors in windows easily:  https://github.com/Akascape/py-window-styles, but only the themes are supported in windows 10. You can try the blur themes if you wish, it works in both win 10 and 11.

Other than this, there is a ready made window template made with overrideredirect method: https://github.com/TomSchimansky/CustomTkinter/discussions/736

It looks the same as regular window, you can change the header colors easily.

![Screenshot](https://github.com/TomSchimansky/CustomTkinter/assets/89206401/8583ce40-56ac-4fed-9cab-ecdee9325b80)

So choose what you want to do. I recommend using pywinstyles to change header color in windows 11 devices and use the simple dark/white headers available in windows 10 by default in customtkinter.

Or you can make 2 versions of it, one for win 11 using pywinstyles and one for win 10 using the ready made template.

**Comment by Akascape** [comment]
I finally made this project: https://github.com/TomSchimansky/CustomTkinter/discussions/1670

![titlebar](https://github.com/Akascape/CTkMenuBar/assets/89206401/da6cd858-f700-476c-a2f0-93a1c6335a4d)

**Reply by LucianoSaldivia** [reply]
I've been away from this project quite a while now, and i plan to keep it that way probably until December at least. From what i can see here, you've done a great job, thanks for making it! Hope it's useful.

**Comment by IAmNotVeozy** [comment]
i HAVENT tried this, but you can use overridedirect.
customize your titlebar, instead of using the master of root, use the master of your titlebar for the menubar
